src/notebookKernel.ts

```diff
@@ -107,15 +170,27 @@ export class NotebookKernel implements vscode.NotebookKernel {
    * @inheritdoc
    */
   public async executeAllCells(
-    document: vscode.NotebookDocument,
-    token: vscode.CancellationToken,
+    document: vscode.NotebookDocument
   ): Promise<void> {
+    this._documentCancellationTokens.get(document)?.cancel();
+    const source = new vscode.CancellationTokenSource();
+    this._documentCancellationTokens.set(document, source);
```

Comment:
Not something I'd hold up on here since this is mostly just a test implementation, but the Reactive way to do this would be having some property

```
requestCancellation = new Subject<{ document?: NotebookDocument; cell?: NotebookCell }>()
```

and then above switch that out

```diff
-takeUntil(observeCodeEvent(token.onCancellationRequested))
+takeUntil(this.requestCancellation.pipe(filter(r => r.document === doc || r.cell = cell)))
```

When writing reactive code I try to keep all data as streams and avoid manually managing external state or properties as much as possible, which reduces complexity and bugs